proposals/csharp-10.0/lambda-improvements.md

```diff
@@ -0,0 +1,121 @@
+
+A method group has a natural type if the method group contains a single method and the method or reduced extension method has no unbound type parameters.
+
+Lambdas or method groups with natural types can be used as initializers in `var` declarations.
+
+If the lambda or method group has no more than 16 parameters and no return value, and all parameters are passed by value, the natural type will be `delegate void System.Action<P1, ..., Pn>(P1, ..., Pn)` where `P1, ..., Pn` are the lambda parameter types.
+
+If the lambda or method group has no more than 16 parameters and a by-value return type, and all parameters are passed by value, the natural type will be `delegate R System.Func<P1, ..., Pn, R>(P1, ..., Pn)` where `P1, ..., Pn` are the lambda parameter types, and `R` is the lambda return type.
+
+Otherwise the natural type will be a synthesized `internal` anonymous `delegate` type with a signature that matches the lambda or method group.
+
+```csharp
+var f1 = () => default;        // error: no natural type
+var f2 = x => { };             // error: no natural type
+var f3 = x => x;               // error: no natural type
+var f4 = () => 1;              // System.Func<int>
+var f5 = () : string => null;  // System.Func<string>
+```
+
+```csharp
+static void F1() { }
+static void F1<T>(this T t) { }
+static void F2(this string s) { }
+
+var f6 = F1;    // error: multiple methods
+var f7 = "".F1; // System.Action
+var f8 = F2;    // System.Action<string> 
+```
+
+### Anonymous delegate type
+If synthesized delegate types are required, the compiler generates generic anonymous delegate types that are shared across all anonymous delegates in the module that have the same number of parameters and same parameter ref kinds.
+
+The names of the synthesized delegate types and the names of the parameters are unspeakable.
+
+The anonymous delegate types are not co- or contra-variant unlike the delegates constructed from `System.Action<>` and `System.Func<>`.
+
+The natural type of a method group does not include any `modopt()` or `modreq()` on the method group.
+
+`fA` and `fB` share a common generic type `internal delegate ref R <anonymous2><P1>(ref P1 <unnamed>)` below:
+```csharp
+var f9 = (ref int i) => { };    // delegate void <anonymous1><int>(ref int <unnamed>);
+var fA = (ref char c) => ref c; // delegate ref char <anonymous2><char, char>(ref char <unnamed>);
+var fB = (ref int i) => ref i;  // delegate ref int <anonymous2><int, int>(ref int <unnamed>);
+```
+
+_Issue: Is it necessary to use `System.Action<>` and `System.Func<>` for the delegate types or should the compiler always generate an anonymous delegate type? There is an observable difference since the anonymous delegate types are not co- or contra-variant._
+
+## Syntax
+
+```antlr
+lambda_expression
+  : attribute_list* modifier* lambda_parameters (':' type)? '=>' (block | body)
```

Comment:
The `:` does feel like the natural thing to do, but I'm concerned that there's going to be ambiguities with `?:` expressions we have today. Are we certain this is a syntax form that we can support?

Reply:
Might easier to parse if we put it inside the `()`? Like `var a = (: string) => "";`? Looks a bit ugly, but I don't believe it's ambiguous.

Reply:
i hate that form @333fred  :)   

THe potential ambiguity would be something like:

`x ? (y) : ...`

is that one of these new lambdas on the RHS of the `?` or is this a parenthesized expression?  Given that this is a legal start, it might be painful to disambiguate.

--

Note: supplying the return type explicitly is one of hte least interesting parts of this proposal for me.  I wouldn't gate it on that, and i'd be happy to drop it and come up with an independent solution there.

Reply:
> i hate that form @333fred :)

I didn't say I _liked_ my form either :P. I just don't think that `:` postfixes are resolvable.

Reply:
How about `(int x, int y => int)` ?

Reply:
I've recorded the possible ambiguity as an issue for now.

---
In reply to: [579504322](https://github.com/dotnet/csharplang/pull/4451#discussion_r579504322) [](ancestors = 579504322)

Reply:
What about `(int a, int b) -> int`?
We have `->` only for pointers and it makes it more clear - at least to me - that is specifies the return type.

Reply:
Doesn't that end up looking like this?

```C#
var f = (int a, int b) -> int => a + b;
```
VS

```C#
var f = (int a, int b) : int => a + b;
```

Not sure about the -> followed by the =>. It does give me haskell vibes though...

Reply:
In this case I dislike the second "arrow", too. 
Perhaps the arrow could be omitted in case of explicit return type followed by a block syntax - which means that we could write for example:

```
var fkt = (int a, int b) -> int
{
    return a + b;
}
```

For simple expression lambdas the return type is not so important - but for complex lambdas this proposal is a very huge improvement.